pyartifactory/objects/repository.py

```diff
@@ -0,0 +1,137 @@
+from __future__ import annotations
+
+import json
+import logging
+from typing import List, overload
+
+import requests
+from pydantic import ValidationError
+
+from pyartifactory.exception import ArtifactoryError, RepositoryAlreadyExistsError, RepositoryNotFoundError
+from pyartifactory.models import AnyRepository, AnyRepositoryResponse
+from pyartifactory.models.repository import (
+    LocalRepository,
+    LocalRepositoryResponse,
+    RemoteRepository,
+    RemoteRepositoryResponse,
+    SimpleRepository,
+    VirtualRepository,
+    VirtualRepositoryResponse,
+)
+from pyartifactory.objects.object import ArtifactoryObject
+from pyartifactory.utils import custom_encoder
+
+logger = logging.getLogger("pyartifactory")
+
+
+class ArtifactoryRepository(ArtifactoryObject):
+    """Models an artifactory repository."""
+
+    _uri = "repositories"
+
+    # Repositories operations
+    def get_repo(self, repo_name: str) -> AnyRepositoryResponse:
+        """
+        Finds repository in artifactory. Raises an exception if the repo doesn't exist.
+        :param repo_name: Name of the repository to retrieve
+        :return: Either a local, virtual or remote repository
+        """
+        try:
+            response = self._get(f"api/{self._uri}/{repo_name}")
+            try:
+                artifact_info: AnyRepositoryResponse = LocalRepositoryResponse.model_validate(response.json())
+            except ValidationError:
+                try:
+                    artifact_info = VirtualRepositoryResponse.model_validate(response.json())
+                except ValidationError:
+                    artifact_info = RemoteRepositoryResponse.model_validate(response.json())
+            return artifact_info
```

Comment:
Shouldn't this be do by `TypeAdapter` and annotated discriminated unions - like this https://github.com/pydantic/pydantic/issues/7487#issuecomment-1724350548 ? as we have 'rclass' field as discriminator between local/remote/virtual

Reply:
Tbf that solution seems pretty unwieldy and at least as verbose as the current one, so I'd like to defer this to a separate PR.

Especially as it does not look like there is a consensus on this issue upstream (yet)

Reply:
I agree with @sigma67 on this